apis/configwatch.go

```diff
@@ -181,3 +206,153 @@ func MonitorTLS(start time.Time, expiry time.Time, conf *config.Config) {
 		}
 	}
 }
+
+func (s *Server) getClientValidator(helloInfo *tls.ClientHelloInfo) func([][]byte, [][]*x509.Certificate) error {
+
+	log.Debug().Msgf("Inside get client validator - %v", helloInfo.Conn.RemoteAddr())
+	return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+		// Verifying client certs with root ca
+		opts := x509.VerifyOptions{
+			Roots:         s.CaCertPool,
+			CurrentTime:   time.Now(),
+			Intermediates: x509.NewCertPool(),
+			KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+		}
+		_, err := verifiedChains[0][0].Verify(opts)
+		if err != nil {
+			log.Error().Err(err).Msgf("certs verification failed")
+			return err
+		}
+
+		log.Debug().Msgf("validating with SAN match rules - %s", s.SANMatchRules)
+		if len(s.SANMatchRules) == 0 {
+			return nil
+		}
+		for _, dnsName := range verifiedChains[0][0].DNSNames {
+			if !validHostname(dnsName, true) {
+				continue
+			}
+			dnsName = toLowerCaseASCII(dnsName)
+			for _, sanMatchRule := range s.SANMatchRules {
+				if !validHostname(sanMatchRule, false) {
+					continue
```

Comment:
a regex might fail the validHostname() funciton and still be a valid regex.  This will prevent us from hitting the matchHostnamesWithRegexp funciton below.  This condition should be added as part of the matchExactly if statement below.  If it fails, we still do the regex match.